usr/lib/api/policy.c

```diff
@@ -0,0 +1,1728 @@
+        case CKM_SHA224_RSA_PKCS_PSS:
+        case CKM_SHA256_RSA_PKCS_PSS:
+        case CKM_SHA384_RSA_PKCS_PSS:
+        case CKM_SHA512_RSA_PKCS_PSS:
+            if (hashmap_find(pp->allowedmechs,
+                             ((CK_RSA_PKCS_PSS_PARAMS *)mech->pParameter)->hashAlg, NULL) == 0) {
+                TRACE_WARNING("POLICY VIOLATION: PSS hash algorithm not allowed by policy.\n");
+                rv = CKR_FUNCTION_FAILED;
+            } else if (policy_is_mgf_allowed(pp,
+                        ((CK_RSA_PKCS_PSS_PARAMS *)mech->pParameter)->mgf) != CKR_OK) {
+                rv = CKR_FUNCTION_FAILED;
+            }
+            break;
+        case CKM_RSA_PKCS_OAEP:
+            if (hashmap_find(pp->allowedmechs,
+                             ((CK_RSA_PKCS_OAEP_PARAMS *)mech->pParameter)->hashAlg, NULL) == 0) {
+                TRACE_WARNING("POLICY VIOLATION: OAEP hash algorithm not allowed by policy.\n");
+                rv = CKR_FUNCTION_FAILED;
+            } else if (policy_is_mgf_allowed(pp,
+                        ((CK_RSA_PKCS_OAEP_PARAMS *)mech->pParameter)->mgf) != CKR_OK) {
+                rv = CKR_FUNCTION_FAILED;
+            }
+            break;
+        case CKM_ECDH1_DERIVE:
+            if (policy_is_kdf_allowed(pp,
+                                      ((CK_ECDH1_DERIVE_PARAMS *)mech->pParameter)->kdf) != CKR_OK)
+                rv = CKR_FUNCTION_FAILED;
+            break;
+        default:
+            break;
+        }
+    }
+ out:
+    if (rv != CKR_OK && sess)
+        sess->session_info.ulDeviceError = CKR_POLICY_VIOLATION;
+    return rv;
+}
+
+static CK_RV policy_update_mech_info(policy_t p, CK_MECHANISM_TYPE mech,
+                                     CK_MECHANISM_INFO_PTR info)
+{
+    struct policy_private *pp = p->priv;
+    /* Silence spurious maybe-uninitialized warning. */
+    struct objstrength tmp = { 0, 0, CK_TRUE };
+
+    if (pp) {
+        if (hashmap_find(pp->allowedmechs, mech, NULL) == 0)
+            return CKR_MECHANISM_INVALID;
+        switch (mech) {
+            /* POLICY: New CKM */
+        case CKM_AES_CBC:
+        case CKM_AES_CBC_PAD:
+        case CKM_AES_CFB128:
+        case CKM_AES_CFB64:
+        case CKM_AES_CFB8:
+        case CKM_AES_CMAC:
+        case CKM_AES_CMAC_GENERAL:
+        case CKM_AES_CTR:
+        case CKM_AES_ECB:
+        case CKM_AES_GCM:
+        case CKM_AES_KEY_GEN:
+        case CKM_AES_MAC:
+        case CKM_AES_MAC_GENERAL:
+        case CKM_AES_OFB:
+        case CKM_IBM_CMAC:
+        case CKM_SHA1_KEY_DERIVATION:
+        case CKM_SHA224_KEY_DERIVATION:
+        case CKM_SHA256_KEY_DERIVATION:
+        case CKM_SHA384_KEY_DERIVATION:
+        case CKM_SHA512_KEY_DERIVATION:
+        case CKM_SSL3_MASTER_KEY_DERIVE:
+        case CKM_SSL3_PRE_MASTER_KEY_GEN:
+            if (policy_update_symmetric(pp, info, CK_TRUE) != CKR_OK) {
+                TRACE_DEVEL("Mechanism 0x%lx blocked by policy!\n", mech);
```

Comment:
Could the mechanism name be printed here?

Reply:
done